opencast/opencast
# Allow generating subtitles in the background

This patch allows generating subtitles in the backgrould, starting the jobs first and attaching the result later in the workflow.

Unfortunately, looking into the code, I also found a lot of minor issues which I fixed as well, but which were hard to separate from this patch:

- `language-fallback` didn't work
- Defaults for generator enging and Whisper.cpp binary fixed
- Defaults for several operations
- Simplified code

### How to test this patch

- Make sure to install Whisper.cpp. You may need to update the configuration `etc/org.opencastproject.speechtotext.impl.engine.WhisperCppEngine.cfg`. Likely you just have to configure something like:
  - `whispercpp.root.path=/home/lars/dev/whisper.cpp/main`
  - `whispercpp.model=/home/lars/dev/whisper.cpp/models/ggml-tiny.bin`
- Update the `fast.yaml` workflow:
```diff
diff --git a/etc/workflows/fast.yaml b/etc/workflows/fast.yaml
index 7a0494ed46..f526ed0a60 100644
--- a/etc/workflows/fast.yaml
+++ b/etc/workflows/fast.yaml
@@ -25,6 +25,7 @@ configuration_panel_json: |-
     ]
   }]
 operations:
+
   - id: defaults
     description: "Applying default configuration values"
     configurations:
@@ -45,6 +46,14 @@ operations:
       - overwrite: false
       - accept-no-media: false
 
+  - id: speechtotext
+    description: Generates subtitles for video and audio files
+    configurations:
+      - source-flavor: '*/prepared'
+      - target-flavor: captions/source
+      - limit-to-one: true
+      - async: true
+
   - id: encode
     fail-on-error: true
     exception-handler-workflow: "partial-error"
@@ -61,9 +70,7 @@ operations:
     description: "Tag captions for publication"
     configurations:
       - source-flavor: "captions/source"
-      - target-flavor: "captions/preview"
       - target-tags: "+engage-download"
-      - copy: true
 
   - id: image
     if: "${straightToPublishing}"
@@ -110,6 +117,12 @@ operations:
       - target-tags: "engage-download"
       - encoding-profile: "player-slides.http"
 
+  - id: speechtotext-attach
+    description: Generates subtitles for video and audio files
+    configurations:
+      - target-flavor: captions/source
+      - target-tags: engage-download
+
   - id: publish-configure
     exception-handler-workflow: "partial-error"
     description: "Publish to preview publication channel"
```

- Download a video with subtitles like:
  https://github.com/user-attachments/assets/932ef91b-580a-4d11-9717-011b7bbdb142
- Convert this to a wav file:
    ```
    ffmpeg -i github-multiline-suggestion.mp4 github-multiline-suggestion.wav
    ```
- Ingest both files:
    ```
    curl -i -u admin:opencast http://localhost:8080/ingest/addMediaPackage/fast -F 'flavor=presentation/source' -F BODY=@github-multiline-suggestion.mp4 -F flavor=presentation/prepared -F BODY2=@github-multiline-suggestion.wav -F title="I 🖤 Opencast"
    ```

### Your pull request should…
